eurovision/notebooks/story.ipynb

```diff
@@ -2090,7 +2039,269 @@
+    "    both = both.rename({'rank': 'actual_rank'}, axis=1)\n",
+    "    return both[['predicted_rank', 'actual_rank']]\n",
+    "    \n",
+    "def get_spearman(predicted_ranks, actual_ranks):\n",
+    "    both_ranks = join_predictions_and_actual(predicted_ranks, actual_ranks)\n",
+    "    return both_ranks['predicted_rank'].corr(both_ranks['actual_rank'], method='spearman')\n",
+    "    \n",
+    "predictions_baseline = predict_baseline(ranks_without_2022, 2022)\n",
+    "actual = get_actual_ranks(2022)\n",
+    "both = join_predictions_and_actual(predictions_baseline, actual)\n",
+    "\n",
+    "print(f'Spearman correlation coefficient for 2022 is {get_spearman(predictions_baseline, actual):.3f}')\n",
+    "both.head(n=10)"
+   ]
+  },
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The Spearman coefficient is rather like its more famous cousin, the Pearson correlation, but is more suitable for ranked data.\n",
+    "A value of 1 indicates that we have perfectly predicted the actual rankings, and -1 means that we have gotten it horribly wrong!\n",
+    "\n",
+    "In this respect, 0.239 does not look like a particularly impressive score.\n",
+    "However, it should be borne in mind that Eurovision results can vary wildly from year to year, and perhaps a more valuable estimate of the accuracy of the model can be obtained by cross-validation.\n",
+    "\n",
+    "To do this, we train the model $N$ times (where $N = 24$ is the total number of years for which we have data), each time leaving out one year's worth of data, and using the model to predict that year's data."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## NOTE: Slovakia only has one final entry in our dataset, in 1998.\n",
+    "## So, when the model is 'trained' on all data excluding 1998, Slovakia doesn't get a rescaled_rank as there is no data to work with.\n",
+    "## The 1998 predictions therefore don't include Slovakia (the Spearman coefficient is calculated based on all other entrants in 1998).\n",
+    "\n",
+    "all_years = df['year'].unique()\n",
+    "\n",
+    "def get_spearman_baseline_without(year):\n",
+    "    rescaled_ranks = train_baseline(excluded_years=[year])\n",
+    "    predictions_baseline = predict_baseline(rescaled_ranks, year)\n",
+    "    actual = get_actual_ranks(year)\n",
+    "    # Remove Slovakia from 1998\n",
+    "    if year == 1998:\n",
+    "        predictions_baseline = predictions_baseline.query('to_code2 != \"SK\"')\n",
+    "        actual = actual.query('to_code2 != \"SK\"')\n",
+    "    return(get_spearman(predictions_baseline, actual))\n",
+    "    \n",
+    "spearmans = [get_spearman_baseline_without(year) for year in all_years]\n",
+    "\n",
+    "print(f'Mean Spearman coefficient across {len(all_years)} years: {np.mean(spearmans)}')\n",
```

Comment:
We discussed this on the call about whether taking the mean makes sense here. I'll leave that to you to figure out, but just flagging it as a reminder.

Reply:
Still not sure about this. Let's see what happens with the other two models and go from there.